src/NServiceBus.TransportTests/When_an_operation_is_cancelled_during_message_processing.cs

```diff
@@ -0,0 +1,47 @@
+﻿namespace NServiceBus.TransportTests
+{
+    using System;
+    using System.Threading.Tasks;
+    using NUnit.Framework;
+    using Transport;
+
+    public class When_an_operation_is_cancelled_during_message_processing : NServiceBusTransportTest
+    {
+        [TestCase(TransportTransactionMode.None)]
+        [TestCase(TransportTransactionMode.ReceiveOnly)]
+        [TestCase(TransportTransactionMode.SendsAtomicWithReceive)]
+        [TestCase(TransportTransactionMode.TransactionScope)]
+        public async Task Should_invoke_recoverability(TransportTransactionMode transactionMode)
+        {
+            var messageProcessingStarted = new TaskCompletionSource<bool>();
+            var recoverabilityInvoked = false;
+
+            OnTestTimeout(() => messageProcessingStarted.SetCanceled());
+
+            await StartPump(
+                (context, _) =>
+                {
+                    messageProcessingStarted.SetResult(true);
+
+                    throw new OperationCanceledException();
+                },
+                (_, __) =>
+                {
+                    recoverabilityInvoked = true;
+
+                    return Task.FromResult(ErrorHandleResult.Handled);
+                },
+                transactionMode);
+
+            await SendMessage(InputQueueName);
+
+            _ = await messageProcessingStarted.Task;
+
+            await Task.Delay(TimeSpan.FromSeconds(1));
```

Comment:
Why do we need to delay here?

Reply:
Because otherwise the exact race condition we identified causes failures. I.e. we throw the exception, we cancel at the same time, and then recoverability does not kick in because `messageProcessingCancellationToken.IsCancellationRequested` is `true`.

Reply:
Alternatively, we could do something like this, but then the test is a little cryptic, since there is no explicit assertion. It will just timeout if the test fails.

```c#
            var messageProcessingStarted = new TaskCompletionSource<bool>();
            var recoverabilityStarted = new TaskCompletionSource<bool>();

            OnTestTimeout(() =>
            {
                messageProcessingStarted.SetCanceled();
                recoverabilityStarted.SetCanceled();
            });

            await StartPump(
                (context, _) =>
                {
                    messageProcessingStarted.SetResult(true);

                    throw new OperationCanceledException();
                },
                (_, __) =>
                {
                    recoverabilityStarted.SetResult(true);

                    return Task.FromResult(ErrorHandleResult.Handled);
                },
                transactionMode);

            await SendMessage(InputQueueName);

            _ = await messageProcessingStarted.Task;
            _ = await recoverabilityStarted.Task;
```